tmail-backend/tmail-third-party/ai-bot/src/main/java/com/linagora/tmail/mailet/rag/LatestEmailReplyExtractor.java

```diff
@@ -0,0 +1,74 @@
+/********************************************************************
+ *  As a subpart of Twake Mail, this file is edited by Linagora.    *
+ *                                                                  *
+ *  https://twake-mail.com/                                         *
+ *  https://linagora.com                                            *
+ *                                                                  *
+ *  This file is subject to The Affero Gnu Public License           *
+ *  version 3.                                                      *
+ *                                                                  *
+ *  https://www.gnu.org/licenses/agpl-3.0.en.html                   *
+ *                                                                  *
+ *  This program is distributed in the hope that it will be         *
+ *  useful, but WITHOUT ANY WARRANTY; without even the implied      *
+ *  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR         *
+ *  PURPOSE. See the GNU Affero General Public License for          *
+ *  more details.                                                   *
+ ********************************************************************/
+package com.linagora.tmail.mailet.rag;
+
+import java.util.regex.Pattern;
+
+import com.google.common.base.Splitter;
+
+public interface LatestEmailReplyExtractor {
+
+    final class RegexBased implements LatestEmailReplyExtractor {
+
+        private static final Pattern QUOTE_PATTERNS = Pattern.compile(
+            "(?mi)" +
+                "^\\s*>.*$|" +
+                "^On .+wrote:.*$|" +
+                "^Le .+a écrit.*$|" +
+                "^De :.*$|" +
+                "^Envoyé :.*$|" +
+                "^À :.*$|" +
+                "^Objet :.*$|" +
+                "^From:.*$|" +
+                "^Sent:.*$|" +
+                "^To:.*$|" +
+                "^Subject:.*$|" +
+                "^On .+from .+<.*>.*$|" +
+                "^Le .+de .+<.*>.*$|" +
+                "^On .+, at .+<.*> wrote:.*$|" +
+                "^Le .+à .+<.*> a écrit.*$"
+        );
+
+        @Override
+        public String cleanQuotedContent(String content) {
+            if (content == null || content.trim().isEmpty()) {
+                return "";
+            }
+            String normalized = content.replace("\r\n", "\n").replace("\r", "\n");
+
+            StringBuilder result = new StringBuilder();
+            for (String line : Splitter.on('\n').split(normalized)) {
+                if (QUOTE_PATTERNS.matcher(line).find()) {
+                    break;
+                }
+                if (line.trim().isEmpty() && result.length() == 0) {
+                    continue;
+                }
+                result.append((line).strip()).append("\n");
+            }
+
+            while (result.length() > 0 && result.charAt(result.length() - 1) == '\n') {
+                result.setLength(result.length() - 1);
+            }
```

Comment:
I do not understand what this block of code does.

Please provide an explanation and think about how we could make the intent clearer to the reader.